internal/controller/devicediscovery_controller.go

```diff
@@ -0,0 +1,235 @@
+		if o.UID == owner.GetUID() {
+			return true
+		}
+	}
+	return false
+}
+
+// reconcile reconciles the devices on the host by comparing the observed devices with the existing NicDevice custom resources (CRs).
+// It deletes CRs that do not represent observed devices, updates the CRs if the status of the device changes,
+// and creates new CRs for devices that do not have a CR representation.
+func (d *DeviceDiscovery) reconcile(ctx context.Context) error {
+	observedDevices, err := d.hostManager.DiscoverNicDevices()
+	if err != nil {
+		return err
+	}
+
+	list := &v1alpha1.NicDeviceList{}
+
+	selectorFields := fields.OneTermEqualSelector("status.node", d.nodeName)
+
+	err = d.Client.List(ctx, list, &client.ListOptions{FieldSelector: selectorFields})
+	if err != nil {
+		log.Log.Error(err, "failed to list NicDevice CRs")
+		return err
+	}
+
+	node := &v1.Node{}
+	err = d.Client.Get(ctx, types.NamespacedName{Name: d.nodeName}, node)
+	if err != nil {
+		log.Log.Error(err, "failed to get node object")
+		return err
+	}
+
+	for _, nicDeviceCR := range list.Items {
+		observedDeviceStatus, exists := observedDevices[nicDeviceCR.Status.SerialNumber]
+
+		if !exists {
+			// Need to delete this CR, it doesn't represent the observedDevice on host anymore
+			err = d.Client.Delete(ctx, &nicDeviceCR)
+			if err != nil {
+				log.Log.Error(err, "failed  to delete NicDevice CR", "device", nicDeviceCR.Name)
+			}
+
+			continue
+		}
+
+		// Need to nullify conditions for deep equal
+		observedDeviceStatus.Conditions = nicDeviceCR.Status.Conditions
+
+		if !reflect.DeepEqual(nicDeviceCR.Status, observedDeviceStatus) {
+			// Status of the device changes, need to update the CR
+			nicDeviceCR.Status = observedDeviceStatus
+
+			err := d.Client.Status().Update(ctx, &nicDeviceCR)
+			if err != nil {
+				log.Log.Error(err, "failed to update NicDevice CR status", "device", nicDeviceCR.Name)
+			}
+		}
+
+		if !hasOwnerRef(node, &nicDeviceCR) {
+			if err := setOwnerRef(ctx, d.Client, node, &nicDeviceCR); err != nil {
+				log.Log.Error(err, "failed to set owner reference for device", "device", nicDeviceCR)
+			}
+		}
+
+		// Device was processed, cleaning it from the map
+		delete(observedDevices, nicDeviceCR.Status.SerialNumber)
+	}
+
+	// Remaining devices don't have CR representation, need to create a CR
+	for _, deviceStatus := range observedDevices {
+		deviceName := d.getCRName(deviceStatus.Type, deviceStatus.SerialNumber)
+		device := &v1alpha1.NicDevice{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      deviceName,
+				Namespace: d.namespace,
+			},
+		}
+
+		err = d.Client.Create(ctx, device)
+
+		if err := setOwnerRef(ctx, d.Client, node, device); err != nil {
+			log.Log.Error(err, "failed to set owner reference for device", "device", device)
+			continue
+		}
+
+		if apierrors.IsAlreadyExists(err) {
+			// Device already exists but was not matched by SerialNumber, which means the status was not applied properly
```

Comment:
how can this happen ?

Reply:
If the device gets created but subsequent status update fails. We can't combine those two in the same API call. It's really a corner case but still can happen

Reply:
that means device serial number is empty -> it will get deleted  at L97

up2u if you want to keep this logic, im probably missing something here.

Reply:
Not really. We list devices based on their `node` field in the status. If the status is empty, the cr will not match the selector. Thus, we try to recover a device without status